frame/tips/src/lib.rs

```diff
@@ -0,0 +1,572 @@
+//! - `retract_tip` - Retract a previous (finders fee registered) report.
+//! - `tip_new` - Report an item worthy of a tip and declare a specific amount to tip.
+//! - `tip` - Declare or redeclare an amount to tip for a particular reason.
+//! - `close_tip` - Close and pay out a tip.
+//!
+//! ## GenesisConfig
+//!
+//! None
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+mod tests;
+mod benchmarking;
+pub mod weights;
+
+use sp_std::if_std;
+
+use sp_std::prelude::*;
+use frame_support::{decl_module, decl_storage, decl_event, ensure, decl_error, Parameter};
+use frame_support::traits::{
+	Currency, Get, ExistenceRequirement::{KeepAlive},
+	ReservableCurrency
+};
+
+use sp_runtime::{ Percent, RuntimeDebug, traits::{
+	Zero, AccountIdConversion, Hash, BadOrigin
+}};
+
+use frame_support::traits::{Contains, ContainsLengthBound};
+use codec::{Encode, Decode};
+use frame_system::{self as system, ensure_signed};
+pub use weights::WeightInfo;
+
+pub type BalanceOf<T> = pallet_treasury::BalanceOf<T>;
+pub type NegativeImbalanceOf<T> = pallet_treasury::NegativeImbalanceOf<T>;
+
+pub trait Trait: frame_system::Trait + pallet_treasury::Trait {
+
+	/// Origin from which tippers must come.
+	///
+	/// `ContainsLengthBound::max_len` must be cost free (i.e. no storage read or heavy operation).
+	type Tippers: Contains<Self::AccountId> + ContainsLengthBound;
+
+	/// The period for which a tip remains open after is has achieved threshold tippers.
+	type TipCountdown: Get<Self::BlockNumber>;
+
+	/// The percent of the final tip which goes to the original reporter of the tip.
+	type TipFindersFee: Get<Percent>;
+
+	/// The amount held on deposit for placing a tip report.
+	type TipReportDepositBase: Get<BalanceOf<Self>>;
+
+	/// The overarching event type.
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+
+	/// Weight information for extrinsics in this pallet.
+	type WeightInfo: WeightInfo;
+}
+
+/// An open tipping "motion". Retains all details of a tip including information on the finder
+/// and the members who have voted.
+#[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug)]
+pub struct OpenTip<
+	AccountId: Parameter,
+	Balance: Parameter,
+	BlockNumber: Parameter,
+	Hash: Parameter,
+> {
+	/// The hash of the reason for the tip. The reason should be a human-readable UTF-8 encoded string. A URL would be
+	/// sensible.
+	reason: Hash,
+	/// The account to be tipped.
+	who: AccountId,
+	/// The account who began this tip.
+	finder: AccountId,
+	/// The amount held on deposit for this tip.
+	deposit: Balance,
+	/// The block number at which this tip will close if `Some`. If `None`, then no closing is
+	/// scheduled.
+	closes: Option<BlockNumber>,
+	/// The members who have voted for this tip. Sorted by AccountId.
+	tips: Vec<(AccountId, Balance)>,
+	/// Whether this tip should result in the finder taking a fee.
+	finders_fee: bool,
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Treasury {
```

Comment:
comment to explain why this is `Treasury` instead of `Tips`

Reply:
Thanks for comment @xlc, added the comment describing the reasons.